### ISSUE TITLE
The "Lazy Lens"

### ISSUE THREAD

**Issue by liscio** [ReactiveCocoa/ReactiveSwift#239]
I apologize for the length of this one, but I feel that this construct I built is very important to add to ReactiveSwift.

I built this extension for `SignalProducer` a while ago while working on the latest [Capo touch](http://capoapp.com/ios) update, and I feel it is an essential addition to the core library. Here's the code for it:

``` swift
extension SignalProducer {
    /// Initializes this SignalProducer with a closure that supplies its value
    /// during the .start event. This initializer is similar to `init(value:)`, 
    /// but it exists to avoid letting "deceptively simple" accessors from being
    /// evaluated immediately from within the body of a `.flatMap` call, e.g.:
    /// 
    /// ```
    /// let imageProducer = underlyingSignal.flatMap(.latest) {
    ///     return SignalProducer(value: $0.slowImageLoad)
    ///         .start(on: backgroundScheduler)
    /// }
    /// ```
    ///
    /// The above code will be evaluated immediately at the call site, and not
    /// on the `backgroundScheduler` as stated. Instead, this pattern can now
    /// work as expected:
    ///
    /// ```
    /// let imageProducer = underlyingSignal.flatMap(.latest) {
    ///     return SignalProducer(lazyValue: { $0.slowImageLoad })
    ///         .start(on: backgroundScheduler)
    /// }
    /// ```
    init(lazyValue: @escaping () -> Value) {
        self.init { observer, disposable in
            observer.send(value: lazyValue())
            observer.sendCompleted()
        }
    }
}

extension PropertyProtocol {
    func lazyLens<T>(getter: @escaping (Value) -> T) -> SignalProducer<T, NoError> {
        return producer.flatMap(.latest) { model in SignalProducer(lazyValue: { getter(model) }) }
    }

    func lazyLens<T>(on scheduler: SchedulerProtocol, getter: @escaping (Value) -> T) -> SignalProducer<T, NoError> {
        return producer.flatMap(.latest) { model in SignalProducer(lazyValue: { getter(model) }).start(on: scheduler) }
    }
}
```

And here's how I use it in practice:

``` swift
final class AlbumViewModel : ViewModelProtocol {
    let title: SignalProducer<String, NoError>
    let albumArtist: SignalProducer<String, NoError>
    let trackText: SignalProducer<String, NoError>
    let artwork: SignalProducer<UIImage?, NoError>
    let blurredArtwork: SignalProducer<UIImage?, NoError>

    let model = MutableProperty<AlbumType?>(nil)
    
    init() {
        title = model.lazyLens { $0?.title ?? "" }
        albumArtist = model.lazyLens { $0?.albumArtist ?? "" }
        trackText = model.lazyLens { $0?.trackCount ?? 0 }.map(AlbumViewModel.textForTrackCount)
        artwork = model.lazyLens(on: artworkScheduler) { $0?.artwork }.replayLazily(upTo: 1)
        blurredArtwork = artwork.flatMap(.latest) { image in return SignalProducer<UIImage?, NoError>(lazyValue: { image?.applyLightEffect() }) }
    }
    
    private let artworkScheduler = QueueScheduler(qos: .userInitiated)

    fileprivate static func textForTrackCount(_ count: Int) -> String {
        if count == 1 {
            return NSLocalizedString( "1 Song", comment: "Album detail text, singular" )
        } else {
            return String.localizedStringWithFormat(NSLocalizedString( "%d Songs", comment: "Album detail text, multiple"), count)
        }
    }
}
```

#### So, what does `lazyLens` do?!

I'm a firm believer in the `ViewModel` pattern where a `MutableProperty` stores a reference to an underlying model object, and the properties of that `ViewModel` are exposed as `lens`es on the underlying model object. They're filtered or otherwise manipulated/calculated to result in the values displayed by the view(controller).

#### But we can already do this with `flatMap(.latest)`, right?

You can create these "lenses" already using the following construct:

```title = model.flatMap(.latest) { SignalProducer(value: $0?.title ?? "") }```

But unfortunately, when the model objects involve image calculations, network or database fetches, or other underlying lazy properties, you are undoing those benefits of being lazy with the above code.

#### But, how is my lazy property no longer lazy?

Because the value is grabbed/calculated immediately when the inner `SignalProducer` is constructed, which is every time `model.value` is set as in my `AlbumViewModel` example above. 

When you write `someViewModel.model.value = newAlbumValue` at the point where you're setting up a `UI{Table,Collection}ViewCell` for display (or before that time, in a pre-fetch of the cell) then there is a **storm** of activity where all these inner `SignalProducer`s are constructed, and the underlying `value:` parameter forces the value be fetched/calculated/constructed immediately to construct the producer.

**This is a very subtle thing to understand here.** It *appears* as if you're not sending the value along to the new `SignalProducer` instance until it is started when you read the code for `SignalProducer.init(value:)`. Unfortunately that's not the case—the value must be calculated right there at initialization time rather than when the inner producer is started.

So that's what `lazyLens` and `init(lazyValue: )` do together—they defer that value calculation until the point where the value is required from the SignalProducer.

#### Does this actually help in practice?

Yes, immensely. Making this change to all my `ViewModel` objects has resulted in the difference between a frequently-pausing table view and a buttery-smooth one that is capable of scrolling over thousands of items at a time.

The benefit comes from the fact that when a `ViewModel` is updated frequently with a new `model` value (as the user is flying by with the scroll), then all those underlying lazy value fetches are never invoked, because those producers are canceled immediately.

Also, this allows the scheduling of those fetches on other queues to actually happen on other queues. For instance, consider this seemingly-harmless line:

``` swift
artwork = model.flatMap(.latest) { SignalProducer(value: $0?.artwork).observe(on: artworkScheduler) }
```

Guess what? Scheduling on the alternate background thread **doesn't work** here. The `artwork` property on the model object that fetches the image (and maybe resizes it) is immediately fetched at the call site in order to pass the parameter into `value:`. Again, **very subtle**. The same issue is true for trying to invoke one of these `SignalProducer`s using `start(on:)`.

So I'd go as far as saying that **`lazyLens` and `init(lazyValue:)` are required to make ReactiveSwift work as intended** when nesting models in this way.

#### Aren't you doing it wrong? Why not make a `SignalProducer` down in the model, or do the fetch in the `ViewModel`?

I've asked myself this a few times, but have ultimately come to realize that `Signal`s do not belong in the model layer.

Whether an operation happens immediately or at a later time should not be exposed by a model object. Instead, I feel that those calculations and related model logic should be built and tested in a synchronous fashion, and the deferring of the calculations should be done in a different layer (such as the `ViewModel`).

Why? It makes development and testing of your business logic straightforward. A `Person` should not give me a callback or promise when I ask for their first name, regardless of how it is stored underneath. ("What's your name?" "I'll tell you later…")

#### Where's the PR?

~~~Coming Soon™ (I've yet to come up with meaningful tests for this.)~~~
See #240

#### What are you looking for feedback on?

Primarily, the naming (surprised?), but also the approach.

I am strong on the concept of a `Lens` for what this construct is giving us. It's a concise concept for exposing "values over time" outlets on an underlying, changing model object.

But it's the "lazy" part I'm expecting to be challenged on. E.g. why not just make the `SignalProducer(value: )` constructor always lazy?

The reason I didn't take that approach is because it means we always have to supply a closure which isn't ideal for simple, flat values. Also, an `@autoclosure` didn't work in practice to clean this up. The lack of an explicit, escaping closure got turned into the equivalent of not evaluating lazily by the compiler.

So that really just leaves `lazyLens` vs `lens`. I'd personally rather just use the `lens` naming and hide the fact that it's invoked lazily. The `lens` effectively serves as a shorthand for `.flatMap(.latest)` and makes it easier on developers to adopt this model. 

I don't see much value in offering a non-lazy variant, so it's probably safe to drop the modifier.

#### Thoughts?

**Comment by liscio** [comment]
Side note: The example code above makes reference to a `ViewModelProtocol`. There's not much to it, so here it is: 

``` swift
protocol ViewModelProtocol {
    associatedtype Model
    var model: MutableProperty<Model?> { get }
}

extension ViewModelProtocol {
    func bind(to model: Model) {
        self.model.value = model
    }
}
```

Looks pretty minimal, but if you build a generic data source for a table or collection view, then you can easily combine the above to create a default implementation for configuring your cells. Something like:

``` swift
protocol ReactiveBindingViewType {
    associatedtype ViewModel
    var viewModel: ViewModel { get }
}

// Imagine a MyFancyDataSource with Cell and Model associatedtypes
extension MyFancyDataSource where Cell: ReactiveBindingViewType, Cell.ViewModel: ViewModelProtocol, Cell.ViewModel == Model {
  func configure(cell: Cell, at indexPath: IndexPath) {
    cell.viewModel.bind(to: self.item(at: indexPath))
  }
}
```

I don't *quite* use it in this way myself, but it's been handy.

**Comment by iv-mexx** [comment]
Sorry, this is not directly related to this proposal, but more general about your architecture.
At first I did not fully understand how you were using `AlbumViewModel`, but as I was writing this, I began to understand. 

* Your `UITableViewCell` / `UICollectionViewCell` has a `AlbumViewModel` (as per the `ReactiveBindingViewType` protocol).
* When a cell is reused, you just update the ViewModel's `model` property (via `bind` as per the `ViewModelProtocol`).
* I guess the cell's attributes are bound via `self.titleLabel.reactive.text <~ self.viewModel.title` once at the setup of the cell
* If the user scrolls so fast that a new viewModel is bound to a specific cell even before the (lazy) values of the previous viewModel have been evaluated, these are cancelled 

Is this summary correct?

If so, I've learned something new and now I'm feeling like I've been doing this wrong the whole time 😆 
Thanks!

**Comment by iv-mexx** [comment]
Just an Idea:

I would expect there to be something similar to e.g. [`Property.map`](https://github.com/ReactiveCocoa/ReactiveSwift/blob/master/Sources/Property.swift#L79)

Would it somehow be possible to implement `lazy` with a Signature like this:  

```
extension PropertyProtocol {
    func lazy() -> Property<Value> {
        // ???
    }

    func lazy(on scheduler: SchedulerProtocol) -> Property<Value> {
        // ???
    }
}
```

I would expect to use it in the ViewModel like this:

```
final class AlbumViewModel : ViewModelProtocol {
    let title: Property<String>
    let albumArtist: Property<String>
    let trackText: Property<String>
    let artwork: Property<UIImage?>
    let blurredArtwork: Property<UIImage?>

    let model = MutableProperty<AlbumType?>(nil)
    
    init() {
        title = model.lazy().map { $0?.title ?? "" }
        albumArtist = model.lazy().map { $0?.albumArtist ?? "" }
        trackText = model.lazy().map { $0?.trackCount ?? 0 }.map(AlbumViewModel.textForTrackCount)
        artwork = model.lazy(on: artworkScheduler).map { $0?.artwork }.replayLazily(upTo: 1)
        blurredArtwork = artwork.flatMap(.latest) { image in return SignalProducer<UIImage?, NoError>(lazyValue: { image?.applyLightEffect() }) }
    }
    
    private let artworkScheduler = QueueScheduler(qos: .userInitiated)

    fileprivate static func textForTrackCount(_ count: Int) -> String {
        if count == 1 {
            return NSLocalizedString( "1 Song", comment: "Album detail text, singular" )
        } else {
            return String.localizedStringWithFormat(NSLocalizedString( "%d Songs", comment: "Album detail text, multiple"), count)
        }
    }
}
```

**Comment by mdiep** [comment]
I think the need here is a valid one. I think we should make it easy to turn non-reactive, synchronous code into reactive, asynchronous code.

But I'm not sold on the specifics of the API. It seems like the mechanism  for this should be exposed as an operator on `Signal` and `SignalProducer`. Then we can also add an operator to `Property`.

I'm curious what you think of that and how you'd update the proposal to reflect that. I have another thought, but I want to see what you'd do there without the influence of it. 😄

**Comment by liscio** [comment]
@iv-mexx Yes, your interpretation is correct. Here's an excerpt where the UI binding is actually done. The `HeaderTableViewCell` is a very simple superclass that only really defines the `@IBOutlet`s.

``` swift
class AlbumItemTableViewCell: HeaderTableViewCell, ReactiveBindingViewType {
    var viewModel = AlbumViewModel()

    override static var nib: UINib? {
        return UINib(nibName: "AlbumItemCell", bundle: nil)
    }

    override func awakeFromNib() {
        super.awakeFromNib()

        let resetImage = SignalProducer(reactive.trigger(for: #selector(self.prepareForReuse)).map { nil as UIImage? })

        primaryLabel.reactive.text <~ viewModel.title
        secondaryLabel.reactive.text <~ viewModel.albumArtist
        tertiaryLabel.reactive.text <~ viewModel.trackText
        albumImageView.reactive.image <~ SignalProducer([resetImage, viewModel.artwork]).flatten(.merge)

        if let blurImageView = self.backgroundView as? UIImageView {
            blurImageView.reactive.image <~ SignalProducer([resetImage, viewModel.blurredArtwork]).flatten(.merge)
        }
    }
}
```

So there's almost nothing in the view/cell classes themselves, as you can see. The only "trickiness" here is my use of a `resetImage` signal that clears the `UIImageView` when `prepareForReuse` is called.

Regarding your question about a `lazy()` operator, it's not really the same because the value evaluation must be performed when the signal is started.

**Comment by liscio** [comment]
> I think the need here is a valid one. I think we should make it easy to turn non-reactive, synchronous code into reactive, asynchronous code.

That's the primary motivation to use RAS to begin with. For me, anyway. :)

> But I'm not sold on the specifics of the API. It seems like the mechanism for this should be exposed as an operator on Signal and SignalProducer. Then we can also add an operator to Property.

I suppose this could be grafted onto `Signal` (or perhaps `SignalProtocol`?) as well, since it's really just a wrapper for a specialized case of `flatMap(.latest)`, anyway.

The only remaining question I have is whether it's equally as valuable in those scenarios, and whether there'd be some (minor) harm in adding confusion. Specifically, throwing `lens` out there onto the `Signal` family might cause folks to misuse (or over-use) it where they could/should be using `map` instead.

But maybe it's not such a bad thing, in reality. Best practices will (eventually) emerge…

> I'm curious what you think of that and how you'd update the proposal to reflect that. I have another thought, but I want to see what you'd do there without the influence of it. 😄

I think it's a fine idea, but I'm curious about what your other thought is.

Also, what are your thoughts on using just the name `lens`, as I implemented it in the PR? The `lazy` aspect is irrelevant outside of the implementation, and I think that it differs from a straight-up `map` operation because this is not really a transformation "in spirit."

**Comment by mdiep** [comment]
> Also, what are your thoughts on using just the name lens, as I implemented it in the PR? The lazy aspect is irrelevant outside of the implementation, and I think that it differs from a straight-up map operation because this is not really a transformation "in spirit."

I think the name _lens_ will be confusing for people. It adds another concept—and one that is increasingly heard of, but not necessarily understood. And I'm not sure it's a truly valid use without a setter.

While this _is_ different than `map`, it's not necessarily that different from `filterMap` (#232). So I'm wondering if it'd make sense to add this as something like `lazyMap`.

**Comment by liscio** [comment]
> I think the name lens will be confusing for people. It adds another concept—and one that is increasingly heard of, but not necessarily understood.

I agree, though one could argue the same for `flatten` and `map`. But yeah—lens is still a pretty "fringe" concept.

> And I'm not sure it's a truly valid use without a setter.

I added a setter in the associated PR, but it makes most sense when `lens` is attached to the `{Mutable,}Property` pair.

The `lens` terminology definitely works best when it's associated with the `Property` types, as it currently stands right now.

> While this is different than map, it's not necessarily that different from filterMap (#232). So I'm wondering if it'd make sense to add this as something like lazyMap.

`lazyMap` might work here as an alternate name, but then it re-raises the question that @iv-mexx had about whether `lazy` could be generalized somehow for other uses. I'll see if I can come up with something, but who knows.

**Comment by liscio** [comment]
This idea (kind of) died in favour of more targeted operators for now. #240 was merged with the read side, and #241 may serve as the write side, if it succeeds.

**Comment by andersio** [comment]
I was thinking of possible prefetching support in ReactiveCollections lately. Apparently both https://github.com/RACCommunity/ReactiveCollections/issues/29 and this issue calls for a real lazy property type. I have outlined a crude idea in https://github.com/RACCommunity/ReactiveCollections/issues/29.
